hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/factory/HoodieSparkKeyGeneratorFactory.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.keygen.factory;
+
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieKeyGeneratorException;
+import org.apache.hudi.keygen.BuiltinKeyGenerator;
+import org.apache.hudi.keygen.ComplexKeyGenerator;
+import org.apache.hudi.keygen.CustomKeyGenerator;
+import org.apache.hudi.keygen.GlobalDeleteKeyGenerator;
+import org.apache.hudi.keygen.NonpartitionedKeyGenerator;
+import org.apache.hudi.keygen.SimpleKeyGenerator;
+import org.apache.hudi.keygen.TimestampBasedKeyGenerator;
+import org.apache.hudi.keygen.constant.KeyGeneratorType;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Factory help to create {@link org.apache.hudi.keygen.KeyGenerator}.
+ * <p>
+ * This factory will try {@link HoodieWriteConfig#KEYGENERATOR_CLASS_PROP} firstly, this ensures the class prop
+ * will not be overwritten by {@link KeyGeneratorType}
+ */
+public class HoodieSparkKeyGeneratorFactory {
```

Comment:
can you help me understand, why do we have 2 keyGen factories? 
And I see sparkKeyGenFactory uses AvorKeyGenFactory to intialize keyGen and if its null, proceeds to initialize by its own. whats the case where avro key gen factory will not work, but spark key gen would work?  I could not find much diff between both key gen. may be I am missing something.

Reply:
> can you help me understand, why do we have 2 keyGen factories?
> And I see sparkKeyGenFactory uses AvorKeyGenFactory to intialize keyGen and if its null, proceeds to initialize by its own. whats the case where avro key gen factory will not work, but spark key gen would work? I could not find much diff between both key gen. may be I am missing something.

when using `KeyGeneratorType` to create `KeyGenerator`, different engines may have different implementations.
pay attention to the difference between `org.apache.hudi.keygen.factory.HoodieAvroKeyGeneratorFactory#createKeyGeneratorByType` and `org.apache.hudi.keygen.factory.HoodieSparkKeyGeneratorFactory#createKeyGeneratorByType`

Reply:
I get it now, thanks.